File: code-test/solutions/java/Day005.java
import java.util.*;

public class Day005 {
    public static boolean solution(String s) {
        // TODO: 괄호 문자열이 올바르면 true, 아니면 false를 반환하세요.
        // 예: "([{}])" -> true, "([)]" -> false
        Map<Character, Character> pairs = new HashMap<>();
        pairs.put(')','(');
        pairs.put('}','{');
        pairs.put(']','[');

        Set<Character> open = new HashSet<>();
        open.add('(');
        open.add('{');
        open.add('[');

        char[] cha = s.toCharArray();
        Stack<Character> stack = new Stack<>();

        for(char item : cha) {
            if(open.contains(item)) {
                stack.push(item);
            } else {
                if (stack.isEmpty() || stack.pop() != pairs.get(item)) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: code-test/solutions/python/day_005.py
def solution(s):
    # TODO: 괄호 문자열이 올바르면 True, 아니면 False를 반환하세요.
    # 예: "([{}])" -> True, "([)]" -> False

    pairs = {
        ')' : '(',
        '}' : '{',
        ']' : '['
    }

    open = {'(','{','['}
    stack = []

    for item in s :
        if item in open :
            stack.append(item)
        elif item in pairs:
            if not stack or stack.pop() != pairs[item] :
                return False



    return not stack

